### PR TITLE
[docker] [skycoindev] fixes #1073 fixes #1481 - Adds Dockerfile to build dev image.

### LINKED ISSUE #1073: [docker] skycoin-dev Docker image
Package in two Docker containers the environment needed to get started with Skycoin development . One of them should include command-line tools only . The second one should add GUI tools and IDE . Build versions of the Docker image on top of `alpine` , `debian` , and `nanoserver` base images.

- [x] A script to (optionally) install Skycoin inside volume in dev mode (e.g. include source and vendor deps) when starting from scratch . Instructions in image README.md .
- [x] Text editor (at least `vim` ~in both containers and [VS Code](https://github.com/hackenfreude/docker-vscode/issues/new) [image](https://hub.docker.com/r/hackenfreude/vscode/) or [Jess Frazelle's Dockerfile](https://github.com/jessfraz/dockerfiles/tree/master/vscode)~ see #1434)
- [x] [fugitive](https://github.com/tpope/vim-fugitive) and other useful vim plugins ( @stdevMauricio u know better than myself )
- [x] [delve](https://github.com/derekparker/delve) debugger.
- [x] `gcc` compiler and standard C headers for libskycoin
- [x] [Criterion](https://github.com/Snaipe/Criterion) to test libskycoin . Consider installing via `make install-deps-libc`
- ~Additional packages in alpine (at least) needs [libpthread-stubs]() , [SSL support](https://github.com/google/cadvisor/issues/1131) , [libc6-compat](https://pkgs.alpinelinux.org/package/edge/main/x86_64/libc6-compat) , libanl ? ([no candidate in alpine](https://github.com/paritytech/parity/issues/5794) ?)~ Do not build alpine image.
- ... etc ...
- ~if there is a way to setup testing for Docker containers , please do so .~ see #1433 
- [ ] Install `make docs` deps [Doxygen](www.doxygen.org/) , [Graphviz](https://www.graphviz.org/) , and [pdflatex](www.math.rug.nl/~trentelman/jacob/pdflatex/pdflatex.html) , [moxygen](https://github.com/sourcey/moxygen)
- [ ] [goautomock](https://github.com/ernesto-jimenez/goautomock) needed for generating mock types e.g. for `Gatewayer` interface .

### LINKED ISSUE #1481: Error when running "make test-libc" in image simelotech/skycoin-dev:feature-stdevMauricio-1073
When executing "make test-libc" prompts up error messages like:

```sh
$ cc -o bin/test_libskycoin_shared lib/cgo/tests/*.c -lskycoin                    -lcriterion -lpthread -Ilib/cgo -Iinclude -Ibuild/usr/include -L build/libskycoin -Lbuild/usr/lib 
In file included from lib/cgo/tests/check_cipher.address.c:6:0:
/usr/local/include/criterion/new/assert.h:436:20: fatal error: stream.h: No such file or directory

In file included from lib/cgo/tests/check_cipher.crypto.c:6:0:
.......
In file included from lib/cgo/tests/check_cipher.hash.c:5:0:
.......
```

and some others.

### PR DESCRIPTION
**Pull request by stdevMauricio**
Fixes #1073 . Fixes #1481 .

Changes:
-Adds Dockerfile to build dev image.

Does this change need to mentioned in CHANGELOG.md?
Yes

### CODE REVIEW

File: docker/images/skycoindev/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# skycoin build
+# reference https://github.com/skycoin/skycoin
+FROM golang:1.10.0-stretch
+
+# dirs
+RUN apt-get update 
+RUN apt-get install -y cmake gdbserver gdb vim ctags vim-scripts screen
+RUN git clone --recurse-submodules -j8 https://github.com/Snaipe/Criterion
+RUN cd /go/Criterion && cmake .
+RUN cd /go/Criterion && make install
```
Comment (olemis) [inline docker/images/skycoindev/Dockerfile]: `cmake` and `make install` in a single step

File: docker/images/skycoindev/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+RUN go get -u github.com/derekparker/delve/cmd/dlv
+
+WORKDIR $GOPATH/src/github.com/skycoin/skycoin
+VOLUME $GOPATH/src/
+
+ENV LD_LIBRARY_PATH=/usr/local/lib
```
Comment (olemis) [inline docker/images/skycoindev/Dockerfile]: Maybe a bit more paths should be added in here , and other env vars needed for compilation and running Makefile .

File: docker/images/skycoindev/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# skycoin build
+# reference https://github.com/skycoin/skycoin
+FROM golang:1.10.0-stretch
+
+# dirs
+RUN apt-get update && apt-get install -y cmake gdbserver gdb vim ctags vim-scripts screen
+RUN git clone --recurse-submodules -j8 https://github.com/Snaipe/Criterion
```
Comment (olemis) [inline docker/images/skycoindev/Dockerfile]: Do not clone this here . Criterion has nothing to do with go lang . Maybe it should be built in a separate image and library binaries plus headers copied in expected locations ?

File: docker/images/skycoindev/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+RUN cd /go/Criterion && cmake .  && make install
+RUN go get -u github.com/derekparker/delve/cmd/dlv
+
+WORKDIR $GOPATH/src/github.com/skycoin/skycoin
+VOLUME $GOPATH/src/
```
Comment (olemis) [inline docker/images/skycoindev/Dockerfile]: In my opinion this should be the order of steps

- `VOLUME ...`
- `mkdir -p ...` to ensure folder exists
- `WORKDIR ...`

File: CHANGELOG.md [hunk 5]
```diff
@@ -25,6 +25,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - `-enable-seed-api` option to enable `POST /wallet/seed`
 - Add `"size"` to block API response data (affects `GET /block`, `GET /blocks` and `GET /last_blocks`)
 - Write [specification for skycoin URIs](https://github.com/skycoin/skycoin#uri-specification) (based upon bip21)
+- Adds Dockerfile in docker/images/skycoindev to build a docker image suitable for development.
```
Comment (olemis) [inline CHANGELOG.md]: Wrong path .

File: docker/images/skycoindev-cli/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,61 @@
+# Creates an image for skycoin development
```
Comment (olemis) [inline docker/images/skycoindev-cli/Dockerfile]: I'd recommend to move this `./docker/images/skycoindev-cli/Dockerfile` => `./docker/images/dev-cli/Dockerfile`

File: docker/images/skycoindev-cli/Dockerfile [hunk 7]
```diff
@@ -17,7 +17,7 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
     && rm -rf /var/lib/apt/lists/* 
 
 # Installs Criterion
-RUN git clone --recurse-submodules -j8 https://github.com/Snaipe/Criterion && \
+RUN git clone --recurse-submodules -j8 https://github.com/mlopezqc/Criterion && \
```
Comment (gz-c) [inline docker/images/skycoindev-cli/Dockerfile]: Why the change?

Comment (olemis) [reply docker/images/skycoindev-cli/Dockerfile]: Error after a change in github.com/Snaipe/Criterion see #1481 . I am of the opinion that we should continue using the same repository but checkout an specific version (hash , since this is in `bleeding`)

Comment (olemis) [reply docker/images/skycoindev-cli/Dockerfile]: ... but a separate repository could be ok . @stdevMauricio if this is the way to go then change the repos to be http://github.com/simelo/Criterion .

Comment (stdevMauricio) [reply docker/images/skycoindev-cli/Dockerfile]: Both suggestions done.

File: docker/images/dev-cli/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+RUN wget http://prdownloads.sourceforge.net/swig/swig-3.0.12.tar.gz \
+    && tar xzf swig-3.0.12.tar.gz \
+    && cd swig-3.0.12 \
+    && sh configure \
+    && make && make install \
+    && cd ../; rm -r swig-3.0.12*
+
+# Installs go development tools
+RUN go get -u github.com/derekparker/delve/cmd/dlv  && \
```
Comment (gz-c) [inline docker/images/dev-cli/Dockerfile]: Don't install the individual linters, see the Makefile:

```make
install-linters: ## Install linters
	go get -u github.com/FiloSottile/vendorcheck
	go get -u github.com/alecthomas/gometalinter
	gometalinter --vendored-linters --install
```

`gometalinter` vendors the linters itself, if you install the linters separately you can have problems when using gometalinter

**Comment by gz-c** [comment]
Mention this docker dev file in the readme, as a subsection in the installation section

File: docker/images/dev-cli/Dockerfile (hunk 8 above)
Comment (stdevMauricio) [reply docker/images/dev-cli/Dockerfile]: I install them beforehand because the skycoin source code is not there when the image is built. The only solution I see is to download the source code, run make install-linters and then delete the source code.

Comment (stdevMauricio) [reply docker/images/dev-cli/Dockerfile]: OK, I'll do it that way, but we must be aware that this way of installing linters has been deprecated by gometalinter and it will be removed soon.

File: README.md [hunk 9]
```diff
@@ -164,6 +164,10 @@ $ docker build -f docker/images/mainnet/Dockerfile \
   -t skycoin:v0.23.0 .
 ```
 
+### Development image
+
+[Development image](docker/images/dev-cli/README.md)
```
Comment (gz-c) [inline README.md]: Add brief description of the purpose of this image here

File: docker/images/dev-cli/Dockerfile (hunk 8 above)
Comment (gz-c) [reply docker/images/dev-cli/Dockerfile]: Where does it say that?

Comment (stdevMauricio) [reply docker/images/dev-cli/Dockerfile]: https://github.com/alecthomas/gometalinter/releases/tag/v2.0.0

Comment (olemis) [reply docker/images/dev-cli/Dockerfile]: What shall be done then ?

File: docker/images/dev-cli/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        ctags \
+        vim-scripts \
+        screen \
+        sudo \
+        doxygen \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/* 
+
+# Installs Criterion
+RUN git clone --recurse-submodules -j8 https://github.com/simelo/Criterion && \
```
Comment (olemis) [inline docker/images/dev-cli/Dockerfile]: @stdevMauricio this repository has been migrated onto https://github.com/skycoin/Criterion . Needs update .

File: docker/images/dev-cli/Dockerfile (hunk 8 above)
Comment (gz-c) [reply docker/images/dev-cli/Dockerfile]: We'll remove it if they remove it. Stay consistent with our Makefile

Comment (olemis) [reply docker/images/dev-cli/Dockerfile]: @gz-c this is already implemented in this `Dockerfile`, changes have been applied by @stdevMauricio

File: docker/images/dev-cli/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    && cd swig-3.0.12 \
+    && sh configure \
+    && make && make install \
+    && cd ../; rm -r swig-3.0.12*
+
+# Installs go development tools
+RUN go get -u github.com/derekparker/delve/cmd/dlv  && \
+    go get -u github.com/FiloSottile/vendorcheck && \
+    go get -u github.com/alecthomas/gometalinter && \
+    gometalinter --vendored-linters --install && \
```
Comment (olemis) [inline docker/images/dev-cli/Dockerfile]: @gz-c `gometalinter` installs linters

File: docker/images/dev-cli/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+        gdb \
+        vim \
+        ctags \
+        vim-scripts \
+        screen \
+        sudo \
+        doxygen \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/* 
+
```
Comment (olemis) [inline docker/images/dev-cli/Dockerfile]: @stdevGeekmidget at this point you gotta install `moxygen` . It requires `npm` so you might need to merge in rules found in `Dockerfile`'s for

- ~[buildpack-deps:stretch-curl](https://github.com/docker-library/buildpack-deps/blob/master/stretch/curl/Dockerfile)~ (not needed)

```
# dpkg -l ca-certificates curl netba
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version      Architecture Description
+++-==============-============-============-=================================
ii  ca-certificate 20161130+nmu all          Common CA certificates
ii  curl           7.52.1-5+deb amd64        command line tool for transferrin
ii  dirmngr        2.1.18-8~deb amd64        GNU privacy guard - network certi
ii  gnupg          2.1.18-8~deb amd64        GNU privacy guard - a free PGP re
ii  netbase        5.4          all          Basic TCP/IP networking system
ii  wget           1.18-5+deb9u amd64        retrieves files from the web
```

- ~[buildpack-deps:buildpack-deps:stretch-scm](https://github.com/docker-library/buildpack-deps/blob/master/stretch/scm/Dockerfile)~ (not needed)

```
# dpkg -l git bzr mercurial subversion

Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version      Architecture Description
+++-==============-============-============-=================================
ii  bzr            2.7.0+bzr661 all          easy to use distributed version c
ii  git            1:2.11.0-3+d amd64        fast, scalable, distributed revis
ii  mercurial      4.0-1+deb9u1 amd64        easy-to-use, scalable distributed
ii  subversion     1.9.5-1+deb9 amd64        Advanced version control system
```

- [buildpack-deps:stretch](https://github.com/docker-library/buildpack-deps/blob/master/stretch/Dockerfile)

```
# dpkg -l autoconf \
>         automake \
>         bzip2 \
>         dpkg-dev \
>         file \
>         g++ \
>         gcc \
>         imagemagick \
>         libbz2-dev \
>         libc6-dev \
>         libcurl4-openssl-dev \
>         libdb-dev \
>         libevent-dev \
>         libffi-dev \
>         libgdbm-dev \
>         libgeoip-dev \
>         libglib2.0-dev \
>         libjpeg-dev \
>         libkrb5-dev \
>         liblzma-dev \
>         libmagickcore-dev \
>         libmagickwand-dev \
>         libncurses5-dev \
>         libncursesw5-dev \
>         libpng-dev \
>         libpq-dev \
>         libreadline-dev \
>         libsqlite3-dev \
>         libssl-dev \
>         libtool \
>         libwebp-dev \
>         libxml2-dev \
>         libxslt-dev \
>         libyaml-dev \
>         make \
>         patch \
>         xz-utils \
> zlib1g-dev
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version      Architecture Description
+++-==============-============-============-=================================
un  autoconf       <none>       <none>       (no description available)
un  automake       <none>       <none>       (no description available)
un  bzip2          <none>       <none>       (no description available)
un  dpkg-dev       <none>       <none>       (no description available)
un  file           <none>       <none>       (no description available)
ii  g++            4:6.3.0-4    amd64        GNU C++ compiler
ii  gcc            4:6.3.0-4    amd64        GNU C compiler
ii  libc6-dev:amd6 2.24-11+deb9 amd64        GNU C Library: Development Librar
un  libtool        <none>       <none>       (no description available)
ii  make           4.1-9.1      amd64        utility for directing compilation
un  patch          <none>       <none>       (no description available)
un  xz-utils       <none>       <none>       (no description available)
dpkg-query: no packages found matching imagemagick
dpkg-query: no packages found matching libbz2-dev
dpkg-query: no packages found matching libcurl4-openssl-dev
dpkg-query: no packages found matching libdb-dev
dpkg-query: no packages found matching libevent-dev
dpkg-query: no packages found matching libffi-dev
dpkg-query: no packages found matching libgdbm-dev
dpkg-query: no packages found matching libgeoip-dev
dpkg-query: no packages found matching libglib2.0-dev
dpkg-query: no packages found matching libjpeg-dev
dpkg-query: no packages found matching libkrb5-dev
dpkg-query: no packages found matching liblzma-dev
dpkg-query: no packages found matching libmagickcore-dev
dpkg-query: no packages found matching libmagickwand-dev
dpkg-query: no packages found matching libncurses5-dev
dpkg-query: no packages found matching libncursesw5-dev
dpkg-query: no packages found matching libpng-dev
dpkg-query: no packages found matching libpq-dev
dpkg-query: no packages found matching libreadline-dev
dpkg-query: no packages found matching libsqlite3-dev
dpkg-query: no packages found matching libssl-dev
dpkg-query: no packages found matching libwebp-dev
dpkg-query: no packages found matching libxml2-dev
dpkg-query: no packages found matching libxslt-dev
dpkg-query: no packages found matching libyaml-dev
dpkg-query: no packages found matching zlib1g-dev
```

- [node:10](https://github.com/nodejs/docker-node/blob/master/10/stretch/Dockerfile) definitely needed, but with `UID=2000` for user `node`
- `npm install moxygen -g` per `moxygen guide`

File: docker/images/dev-cli/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    go get -u github.com/zmb3/gogetdoc && \
+    go get -u golang.org/x/tools/cmd/guru && \
+    go get -u github.com/davidrjenni/reftools/cmd/fillstruct && \
+    go get -u github.com/rogpeppe/godef && \
+    go get -u github.com/fatih/motion && \
+    go get -u github.com/nsf/gocode && \
+    go get -u github.com/jstemmer/gotags && \
+    go get -u github.com/josharian/impl && \
+    go get -u github.com/fatih/gomodifytags && \
+    go get -u github.com/dominikh/go-tools/cmd/keyify
```
Comment (olemis) [inline docker/images/dev-cli/Dockerfile]: In here install [goautomock](https://github.com/ernesto-jimenez/goautomock#installation) .

File: docker/images/dev-cli/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+        gdbserver \
+        gdb \
+        vim \
+        ctags \
+        vim-scripts \
+        screen \
+        sudo \
+        doxygen \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
```
Comment (olemis) [inline docker/images/dev-cli/Dockerfile]: @stdevGeekmidget in here install [Graphviz](https://www.graphviz.org/) , and [pdflatex](www.math.rug.nl/~trentelman/jacob/pdflatex/pdflatex.html)

File: docker/images/dev-cli/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    go get -u github.com/rogpeppe/godef && \
+    go get -u github.com/fatih/motion && \
+    go get -u github.com/nsf/gocode && \
+    go get -u github.com/jstemmer/gotags && \
+    go get -u github.com/josharian/impl && \
+    go get -u github.com/fatih/gomodifytags && \
+    go get -u github.com/dominikh/go-tools/cmd/keyify
+
+# Install vim-go development tools
+RUN git clone https://github.com/fatih/vim-go /usr/share/vim/vim80/pack/dev/start/vim-go && \
```
Comment (olemis) [inline docker/images/dev-cli/Dockerfile]: Ensure that these `vim-go` dependencies are installed . I had to install'em by executing `vim-go`'s `:GoInstallBinaries` command.

```
vim-go: gogetdoc not found. Installing github.com/zmb3/gogetdoc to folder $GOPATH/bin/
vim-go: fillstruct not found. Installing github.com/davidrjenni/reftools/cmd/fillstruct to folder $GOPATH/bin/
vim-go: motion not found. Installing github.com/fatih/motion to folder $GOPATH/bin/
vim-go: errcheck not found. Installing github.com/kisielk/errcheck to folder $GOPATH/bin/
vim-go: gotags not found. Installing github.com/jstemmer/gotags to folder $GOPATH/bin/
vim-go: impl not found. Installing github.com/josharian/impl to folder $GOPATH/bin/
vim-go: gomodifytags not found. Installing github.com/fatih/gomodifytags to folder $GOPATH/bin/
vim-go: keyify not found. Installing github.com/dominikh/go-tools/cmd/keyify to folder $GOPATH/bin/
vim-go: asmfmt not found. Installing github.com/klauspost/asmfmt/cmd/asmfmt to folder $GOPATH/bin/
vim-go: gometalinter not found. Installing github.com/alecthomas/gometalinter to folder $GOPATH/bin/
```

Comment (olemis) [reply docker/images/dev-cli/Dockerfile]: Of course this needs to be done from the command line rather than through a `vim` command .

Comment (olemis) [reply docker/images/dev-cli/Dockerfile]: I double checked . The former output was in my laptop . This is what I got inside the container 

```
vim-go: gorename not found. Installing golang.org/x/tools/cmd/gorename to folder
 /go/bin/
vim-go: asmfmt not found. Installing github.com/klauspost/asmfmt/cmd/asmfmt to folder
 /go/bin/
```

File: docker/images/dev-cli/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,183 @@
+# Creates an image for skycoin development
+FROM golang:1.10.0-stretch
```
Comment (gz-c) [inline docker/images/dev-cli/Dockerfile]: golang:1.10.2 ?
